core/mintController.js

```diff
@@ -1,50 +1,57 @@
 import { makePrivateName } from '../util/PrivateName';
 
 export function makeBasicMintController() {
-  // Map from purse or payment to the rights it currently
-  // holds. Rights can move via payments
-
-  // purse/payment to amount
-  let rights = makePrivateName();
-
-  function recordPayment(src, payment, amount, srcNewRightsAmount) {
-    rights.set(src, srcNewRightsAmount);
-    rights.init(payment, amount);
+  // An asset can either be a purse or payment. An asset controller
+  // keeps track of either all of the purses (purseController) or all
+  // of the payments (paymentController).
+  function makeAssetController(type) {
+    // asset to amount
+    let assets = makePrivateName();
+    return {
+      updateAmount(asset, newAmount) {
+        assets.set(asset, newAmount);
+      },
+      recordNew(asset, initialAmount) {
+        assets.init(asset, initialAmount);
+      },
+      getAmount(asset) {
+        return assets.get(asset);
+      },
+      getType() {
+        return `${type}`;
+      },
+      has(asset) {
+        return assets.has(asset);
+      },
+      destroyAll() {
+        assets = makePrivateName(); // reset completely
+      },
+    };
   }
 
+  // destroy is outside of an assetController because it could affect
+  // purses or payments
   function destroy(_amount) {
     throw new Error('destroy is not implemented');
   }
 
-  function destroyAll() {
-    rights = makePrivateName(); // reset rights
-  }
-
-  function recordDeposit(
-    srcPayment,
-    srcNewRightsAmount,
-    purse,
-    purseNewRightsAmount,
-  ) {
-    rights.set(srcPayment, srcNewRightsAmount);
-    rights.set(purse, purseNewRightsAmount);
-  }
+  const purseController = makeAssetController('purse');
+  const paymentController = makeAssetController('payment');
 
-  function recordMint(purse, initialAmount) {
-    rights.init(purse, initialAmount);
+  function isPurse(asset) {
+    return purseController.has(asset);
   }
 
-  function getAmount(pursePayment) {
-    return rights.get(pursePayment);
+  function isPayment(asset) {
+    return paymentController.has(asset);
   }
 
   const mintController = {
```

Comment:
Curious why you're defining the methods by defining functions first and then gathering them into properties. If they call each other, that could be a good reason. Otherwise, my inclination is to define them inside the object using concise method syntax. This is (slightly) semantically better too, in that methods (functions defined using concise method syntax) have no [[Construct]] behavior, i.e., you can't `new` them.

Reply:
Hmm, I think this might be a misunderstanding. All of the methods are defined using concise method syntax. `isPurse()` and `isPayment()` here aren't object methods so they're just defined as functions.

Reply:
I see. My criticism applied to the prior version but I had not appreciated that it is mostly fixed in this one. However, why is `destroy` a function rather than a concise method?

Reply:
Destroy takes in an amount and acts on both purses and payments (potentially in the same call) depending on where the pixels in the amount was located, so there wasn't an easy way to contain it within a purseController or paymentController.

Reply:
Ah, thanks for explaining. I could definitely make this a concise method. Would you like for that change to be part of this PR?

Reply:
y